agent/server.go

```diff
@@ -105,6 +108,35 @@ func (r *server) UpdateTimeline(ctx context.Context, req *pb.UpdateRequest) (*pb
 	return &pb.UpdateResponse{}, nil
 }
 
+// Stop stops the grpc server and any additional http servers.
+// TODO: modify Stop to return error
+func (r *server) Stop() {
+	// TODO: pass context in as a parameter.
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if err := r.stopHTTPServers(ctx); err != nil {
+		log.WithError(err).Error("Some HTTP servers failed to shutdown.")
+	}
+
+	r.Server.Stop()
+}
+
+// stopHTTPServers shuts down all listening http servers.
+func (r *server) stopHTTPServers(ctx context.Context) error {
+	var errors []error
+	for _, srv := range r.httpServers {
+		err := srv.Shutdown(ctx)
+		if utils.IsContextDone(ctx) {
```

Comment:
I was thinking that if the context is done every other call to Shutdown will return immediately and is wasting resources.

I guess we don't need to handle context is done case separately so that we can log each server that failed to shutdown before context is done.